### ISSUE TITLE
System.Net.Security tests failing with: "The credentials supplied to the package were not recognized"

### ISSUE THREAD

**Issue by ianhays** [dotnet/corefx#16516]
There is a group of tests that are failing intermittently with the error ```System.ComponentModel.Win32Exception : The credentials supplied to the package were not recognized```. I'm tracking their failures in this issue as they are likely due to the same underlying issue.

The following tests have failed at some point with this error:
- [SslStream_StreamToStream_Authentication_Success](https://github.com/dotnet/corefx/issues/17103)
- [SslStream_StreamToStream_HandshakeAlert_Ok](https://github.com/dotnet/corefx/issues/12706)
- [SslStream_StreamToStream_DataAfterShutdown_Fail](https://github.com/dotnet/corefx/issues/12706)
- [SslStream_StreamToStream_ServerInitiatedCloseNotify_Ok](https://github.com/dotnet/corefx/issues/12683)
- [SslStream_StreamToStream_ClientInitiatedCloseNotify_Ok](https://github.com/dotnet/corefx/issues/15785)
- [SslStream_StreamToStream_Write_ReadByte_Success](https://github.com/dotnet/corefx/issues/16338)
- [ServerAsyncAuthenticate_AllClientVsIndividualServerSupportedProtocols_Success(serverProtocol: Tls11)](https://github.com/dotnet/corefx/issues/8744)
- [ServerAsyncAuthenticate_AllClientVsIndividualServerSupportedProtocols_Success](https://github.com/dotnet/corefx/issues/16346)
- [ServerAsyncAuthenticate_EachSupportedProtocol_Success(protocol: Tls11)](https://github.com/dotnet/corefx/issues/8744)
- [ServerAsyncAuthenticate_MismatchProtocols_Fails(serverProtocol: Tls, clientProtocol: Tls11, expectedException: typeof(System.Security.Authentication.AuthenticationException))](https://github.com/dotnet/corefx/issues/16651)

None of these test failures are reproducible on local machines regardless of the number of runs executed, which suggests the issue lies in either the infrastructure or interference from other tests run alongside these. They only fail on Windows.

It might be helpful to get another pair of eyes on this to see if anything stands out. As it is, I've been unable to narrow the symptoms down to any definitive problem and @steveharter and my's attempts to reproduce failures have been fruitless. I'll be disabling this entire group of tests to regain stability until we can get more information on their source.

**Comment by Jiayili1** [comment]
This issue is repro on build 20170227.01, detail: https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20170227.01/workItem/System.Net.Security.Tests/analysis/xunit/System.Net.Security.Tests.ServerAsyncAuthenticateTest~2FServerAsyncAuthenticate_AllClientVsIndividualServerSupportedProtocols_Success(serverProtocol:%20Tls)

Failed test:  System.Net.Security.Tests.ServerAsyncAuthenticateTest/ServerAsyncAuthenticate_AllClientVsIndividualServerSupportedProtocols_Success(serverProtocol: Tls)

**Comment by ianhays** [comment]
Not sure if it is interesting, but `ServerAsyncAuthenticate_AllClientVsIndividualServerSupportedProtocols_Success`and `ServerAsyncAuthenticate_EachSupportedProtocol_Success` are **only** failing for `Tls11`, not `Tls1s` or `Tls12`

**Comment by ianhays** [comment]
We hit another one: https://ci.dot.net/job/dotnet_corefx/job/master/job/windows_nt_debug_prtest/4435/testReport/junit/System.Net.Security.Tests/SchSendAuxRecordTest/SslStream_ClientAndServerUsesAuxRecord_Ok/

System.Net.Security.Tests.SchSendAuxRecordTest.SslStream_ClientAndServerUsesAuxRecord_Ok (from (empty))

MESSAGE:
System.ComponentModel.Win32Exception : The credentials supplied to the package were not recognized

**Comment by steveharter** [comment]
@ianhays have you been able to grab a test machine from Jenkins that had a failure? I would like to compare the OS version and updates against my local environment - I've been unable to repro these on local Win7 although I have all updates etc applied. Thanks

**Comment by ianhays** [comment]
Yeah, I tried to repro them on a Jenkins machine before. I just took the machine offline. I'll send you the links and login info so you can play around on it.

**Comment by Jiayili1** [comment]
Failed again on build 20170310.01, detail: https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20170310.01/workItem/System.Net.Security.Tests/analysis/xunit/System.Net.Security.Tests.SslStreamNetworkStreamTest~2FSslStream_SendReceiveOverNetworkStream_Ok

**Comment by ianhays** [comment]
FYI @steveharter the test re-enabled in https://github.com/dotnet/corefx/pull/16877 just failed [again here](https://mc.dot.net/#/product/netcore/master/source/official%7E2Fcorefx%7E2Fmaster%7E2F/type/test%7E2Ffunctional%7E2Fcli%7E2F/build/20170310.01/workItem/System.Net.Security.Tests/analysis/xunit/System.Net.Security.Tests.SslStreamNetworkStreamTest%7E2FSslStream_SendReceiveOverNetworkStream_Ok)

**Comment by Jiayili1** [comment]
This issue is repro on build [20170315.01](https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20170315.01/workItem/System.Net.Security.Tests/analysis/xunit/System.Net.Security.Tests.SslStreamNetworkStreamTest~2FSslStream_SendReceiveOverNetworkStream_Ok)

**Comment by stephentoub** [comment]
Again here:
https://ci.dot.net/job/dotnet_corefx/job/master/job/windows_nt_debug_prtest/4723/consoleText

**Comment by stephentoub** [comment]
Again here:
https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_win10_debug/124/consoleText

**Comment by ianhays** [comment]
Four failures in a week suggests there hasn't been any difference in the failure rate of the test re-enabled and modified in https://github.com/dotnet/corefx/pull/16877. Since we've got enough failure reports for analysis of the test, I'm going to disable it again to clean up the builds.

**Comment by Jiayili1** [comment]
Failed again here: https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_win10_debug/lastCompletedBuild/testReport/System.Net.Security.Tests/SslStreamNetworkStreamTest/SslStream_SendReceiveOverNetworkStream_Ok/

**Comment by CIPop** [comment]
Continuously ran 20 separate processes with a random delay between 0-200 between runs for S.N.Security on my dev station for 16hrs. None of the processes experienced crashes.

**Comment by CIPop** [comment]
For reference, a readable stack is:

```
System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncAuthenticate_MismatchProtocols_Fails(serverProtocol: Tls, clientProtocol: Tls11, expectedException: typeof(System.Security.Authentication.AuthenticationException)) [FAIL]
15:48:14         Assert.Throws() Failure
15:48:14         Expected: typeof(System.Security.Authentication.AuthenticationException)
15:48:14         Actual:   typeof(System.ComponentModel.Win32Exception): The credentials supplied to the package were not recognized
15:48:14         Stack Trace:
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\Common\src\Interop\Windows\sspicli\SSPIWrapper.cs(167,0): at System.Net.SSPIWrapper.AcquireCredentialsHandle(SSPIInterface secModule, String package, CredentialUse intent, SCHANNEL_CRED scc)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslStreamPal.Windows.cs(407,0): at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(CredentialUse credUsage, SCHANNEL_CRED secureCredential)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslStreamPal.Windows.cs(131,0): at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(X509Certificate certificate, SslProtocols protocols, EncryptionPolicy policy, Boolean isServer)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SecureChannel.cs(680,0): at System.Net.Security.SecureChannel.AcquireServerCredentials(Byte[]& thumbPrint)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SecureChannel.cs(783,0): at System.Net.Security.SecureChannel.GenerateToken(Byte[] input, Int32 offset, Int32 count, Byte[]& output)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SecureChannel.cs(704,0): at System.Net.Security.SecureChannel.NextMessage(Byte[] incoming, Int32 offset, Int32 count)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslState.cs(767,0): at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslState.cs(982,0): at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslState.cs(937,0): at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslState.cs(885,0): at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslState.cs(693,0): at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslState.cs(614,0): at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslStream.cs(160,0): at System.Net.Security.SslStream.BeginAuthenticateAsServer(X509Certificate serverCertificate, Boolean clientCertificateRequired, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation, AsyncCallback asyncCallback, Object asyncState)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslStream.cs(257,0): at System.Net.Security.SslStream.<>c__DisplayClass35_0.<AuthenticateAsServerAsync>b__0(AsyncCallback callback, Object state)
15:48:14               at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl(Func`3 beginMethod, Func`2 endFunction, Action`1 endAction, Object state, TaskCreationOptions creationOptions)
15:48:14               at System.Threading.Tasks.TaskFactory.FromAsync(Func`3 beginMethod, Action`1 endMethod, Object state)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\src\System\Net\Security\SslStream.cs(257,0): at System.Net.Security.SslStream.AuthenticateAsServerAsync(X509Certificate serverCertificate, Boolean clientCertificateRequired, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation)
15:48:14            D:\j\workspace\windows_nt_de---4526f5ff\src\System.Net.Security\tests\FunctionalTests\ServerAsyncAuthenticateTest.cs(154,0): at System.Net.Security.Tests.ServerAsyncAuthenticateTest.<ServerAsyncSslHelper>d__11.MoveNext()
15:48:14            --- End of stack trace from previous location where exception was thrown ---
15:48:14               at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
15:48:14               at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
15:48:14               at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

**Comment by steveharter** [comment]
Also no repro after re-enabling the tests and running in loop 5,000 times on Windows 10 and 2008r2

**Comment by steveharter** [comment]
Adding test SslStreamStreamToStreamTest_Sync.SslStream_StreamToStream_Successive_ClientWrite_WithZeroBytes_Success as a failure per #18623

**Comment by CIPop** [comment]
@steveharter @ianhays @karelz @mmitche At this stage I'd like to add a Debug.Fail and enable the tests to collect more information. This will work only if procdump or WinDBG are installed on the CI/debug system (release builds won't be affected). 

How can I collect dumps from the CI machines afterwards?

**Comment by karelz** [comment]
@mellinoe can help with details. Debug.Fail was recently broken btw.

**Comment by mellinoe** [comment]
If we have a recent crash in CI, we should be able to grab the binaries and the crash dump. Unfortunately, it looks like the crash dump web page is down right now... I'll see if we can get it back online.

**Comment by CIPop** [comment]
@mellinoe There are no crashes because all these tests have been disabled. I have two questions:
* Would adding a Debug.Fail be sufficient or should I write my own [MiniDumpWriteDump ](https://msdn.microsoft.com/en-us/library/windows/desktop/ms680360(v=vs.85).aspx) code to write the dmp file? 
* If Debug.Fail is sufficient, will the tool collect a full-memory dump and will I have access to the symbols?

@karelz I've tested Debug.Fail with both WinDBG and Procdump and in both cases I got the right information as long as the dumps are full-memory.

**Comment by mellinoe** [comment]
`Debug.Fail` should terminate the process and create a crash dump which we will upload, but I'm not sure what type of dumps we have the CI machines configured to collect. It may not be a full-memory dump, I think that depends on what we have the registry keys set to.

**Comment by CIPop** [comment]
@mellinoe Do you have access to any dump produced by the CI system? It's easy to tell if it's full or partial.
The default config in Windows is to collect partial dumps which are next to useless for managed code debugging.

**Comment by CIPop** [comment]
@mellinoe answered my questions offline. We should be able to collect full-dumps. 
I'll propose a PR with the necessary changes.

**Comment by CIPop** [comment]
/cc @Priya91 
I was finally able to repro in a debugger but I'm not sure what causes this yet.

This has the same cause as #9543, #15777 and #16167: the `X509Certificate` object is disposed sometime after the `cred.paCred` is set [here](https://github.com/dotnet/corefx/blob/ec2a6190efa743ab600317f44d757433e44e859b/src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs#L381) but before ACH is called [here](https://github.com/dotnet/corefx/blob/ec2a6190efa743ab600317f44d757433e44e859b/src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs#L402). This causes an exception that is caught and ACH is tried again [here](https://github.com/dotnet/corefx/blob/ec2a6190efa743ab600317f44d757433e44e859b/src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs#L407). It will fail the same way. 

@bartonjs @stephentoub @steveharter All tests that sometimes fail use a similar pattern to  [SslStream_StreamToStream_HandshakeAlert_Ok](https://github.com/dotnet/corefx/blob/4b07f700c4df2e24413b91edaa55a621d5221d38/src/System.Net.Security/tests/FunctionalTests/SslStreamAlertsTest.cs#L27).

Here are some of the possibilities that I currently see:

1. The code in [GetCertWithPrivateKey](https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/Common/tests/System/Net/Configuration.Certificates.cs#L37) is incorrect. As @davidsh said before, disposing the X509Certificate2Collection (which goes out of scope in that function) will also close the related handles of all other certificates.

Re this @bartonjs answered:
> X509Certificate2Collection isn't inherently IDisposable; nothing would Dispose a cert unless you did it manually. That said, the X509Certificate tests do hold static data, and each test that wants the data clones it anew.

I'm not sure if you meant that the code is incorrect as written or just stated that each test will clone the static data - which I think is what we want to avoid having other tests executing in parallel disposing the certificate.

2. There's some connection between the certificates and the loop that disposes each cert in the collection is somehow affecting the only certificate that is returned.

3. The `using` statement is behaving differently from what I've assumed when await/async is involved and somehow disposes the object before the continuations.

**Comment by steveharter** [comment]
@CIPop how did you repro this? Which test and can you repro it again?

If you are able to repro it at will, can you also repro with just that test alone? To eliminate any thought of potential interference. 

How do you know the X509Certificate was disposed -- a zero `secureCredential` handle passed into AcquireCredentialsHandle(...)?

Also on possibility # 1, if the X509Certificate2Collection goes out of scope, that should be fine because we continue to have a reference to the X509Certificate via the `using` statement... But I suppose that's what you meant by # 3.

**Comment by CIPop** [comment]
> @CIPop how did you repro this? Which test and can you repro it again?

In a debugger by disposing the certificate handle right before the ACH call.

> How do you know the X509Certificate was disposed -- a zero secureCredential handle passed into AcquireCredentialsHandle(...)?

This was a wrong assumption: the actual bug is a Windows level race condition during the PFX file import: Importing the same file in parallel from multiple processes / threads may end up corrupting the certificate state.

**Comment by CIPop** [comment]
Please reopen attaching:
1. Exception and test log
2. Crash-dump generated by the Debug.Assert (if this failed in a debug build)

**Comment by karelz** [comment]
@CIPop do you expect the tests to fail?

BTW: Crash-dump has to be pulled manually by dev from Dumpling tool. Exception and test logs should be already part of general test failure reporting, right?

**Comment by ViktorHofer** [comment]
Happened here again: https://ci.dot.net/job/dotnet_corefx/job/master/job/netfx_windows_nt_release_prtest/133/

**Comment by CIPop** [comment]
Thanks for reporting @ViktorHofer . From the stack that's a different problem or the same problem but while calling a different Win32 API. 
This one has been instrumented to crash XUnit and create a dump which didn't happen.

**Comment by CIPop** [comment]
> that's a different problem

Spoke too soon: the failure is observed in a release build and so the instrumentation didn't exist. It's probably just the same exception marshaled on the Xunit thread by the APM mechanism. 

```
      System.Net.Security.Tests.ServerAsyncAuthenticateTest.ServerAsyncAuthenticate_AllClientVsIndividualServerSupportedProtocols_Success(serverProtocol: Tls11) [FAIL]
19:28:50         System.ComponentModel.Win32Exception : The credentials supplied to the package were not recognized
19:28:50         Stack Trace:
19:28:50              at System.Net.Security.SslState.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
19:28:50              at System.Net.Security.SslState.EndProcessAuthentication(IAsyncResult result)
19:28:50              at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
19:28:50           --- End of stack trace from previous location where exception was thrown ---
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
19:28:50           D:\j\workspace\netfx_windows---50c21bdb\src\Common\tests\System\Threading\Tasks\TaskTimeoutExtensions.cs(19,0): at System.Threading.Tasks.TaskTimeoutExtensions.<TimeoutAfter>d__0.MoveNext()
19:28:50           --- End of stack trace from previous location where exception was thrown ---
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
19:28:50           D:\j\workspace\netfx_windows---50c21bdb\src\System.Net.Security\tests\FunctionalTests\ServerAsyncAuthenticateTest.cs(171,0): at System.Net.Security.Tests.ServerAsyncAuthenticateTest.<ServerAsyncSslHelper>d__11.MoveNext()
19:28:50           --- End of stack trace from previous location where exception was thrown ---
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
19:28:50           D:\j\workspace\netfx_windows---50c21bdb\src\System.Net.Security\tests\FunctionalTests\ServerAsyncAuthenticateTest.cs(95,0): at System.Net.Security.Tests.ServerAsyncAuthenticateTest.<ServerAsyncAuthenticate_AllClientVsIndividualServerSupportedProtocols_Success>d__9.MoveNext()
19:28:50           --- End of stack trace from previous location where exception was thrown ---
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
19:28:50           --- End of stack trace from previous location where exception was thrown ---
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
19:28:50           --- End of stack trace from previous location where exception was thrown ---
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
19:28:50              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
19:28:52     Finished:    System.Net.Security.Tests
19:28:52   === TEST EXECUTION SUMMARY ===
19:28:52 D:\j\workspace\netfx_windows---50c21bdb\Tools\tests.targets(338,5): warning :    System.Net.Security.Tests  Total: 109, Errors: 0, Failed: 1, Skipped: 0, Time: 243.960s [D:\j\workspace\netfx_windows---50c21bdb\src\System.Net.Security\tests\FunctionalTests\System.Net.Security.Tests.csproj]
19:28:52   Trying to find crash dumps for project: System.Net.Security.Tests
19:28:52   No new dump file was found in C:\Users\DOTNET~1\AppData\Local\Temp\CoreRunCrashDumps
19:28:52   Finished running tests.  End time=19:28:51.90, Exit code = 0
19:28:52 D:\j\workspace\netfx_windows---50c21bdb\Tools\tests.targets(338,5): warning MSB3073: The command "D:\j\workspace\netfx_windows---50c21bdb\bin/Windows_NT.AnyCPU.Release/System.Net.Security.Tests/netstandard//RunTests.cmd D:\j\workspace\netfx_windows---50c21bdb\bin/testhost/netfx-Windows_NT-Release-x64/" exited with code -1. [D:\j\workspace\netfx_windows---50c21bdb\src\System.Net.Security\tests\FunctionalTests\System.Net.Security.Tests.csproj]
19:28:52 
D:\j\workspace\netfx_windows---50c21bdb\Tools\tests.targets(346,5): error : One or more tests failed while running tests from 'System.Net.Security.Tests' please check D:\j\workspace\netfx_windows---50c21bdb\bin/Windows_NT.AnyCPU.Release/System.Net.Security.Tests/netstandard/testResults.xml for details! [D:\j\workspace\netfx_windows---50c21bdb\src\System.Net.Security\tests\FunctionalTests\System.Net.Security.Tests.csproj]
```

From the name `Windows_NT` intersected with the URIs at https://github.com/dotnet/corefx#build--test-status I'd say this failed on Windows 8.1.
@mmitche please confirm (also, could we change the names to reflect actual OS versions?)

**Comment by CIPop** [comment]
We have another repro in a release build. Still waiting for a DBG build with more actionable data (a full-memory crash dump).

Once you observe such a repro, please use https://dumpling.azurewebsites.net/ or directly at us and the CI team ( @cipop @mmitche @steveharter) to collect the dump and symbols before the machine goes offline.

To collect the dump with full symbol information:
```
		Open WinDBG on the same machine and open the <pathto>.dmp file.
		.symfix
		.srcfix
		.reload /f
               .dumpcab /a <pathto>.cab
```

**Comment by CIPop** [comment]
@DavidGoll @karelz I suggest moving this to 2.1/Future while we're waiting for a perfect repro.

**Comment by bartonjs** [comment]
If these are only crashing in release, it might be a KeepAlive-type problem... running with COMPlus_GCStress=3 seems good.

**Comment by CIPop** [comment]
Thanks @bartonjs! Had no idea that existed... 
Is there more documentation about `COMPlus_GCStress` and other such options?

**Comment by bartonjs** [comment]
Probably, but I'm not finding it easily.  https://github.com/dotnet/coreclr/blob/master/src/vm/eeconfig.h#L654 shows the GCStress flags options.  3 being alloc+transition.

```C#
private static void DoSomething()
{
    // Notice the lack of a `using`
    SafeHandle h = GetSomeSafeHandle();
    DoSomething(h);
    // marker 3
    // And we never speak of h again.
} // marker 1.

private static void DoSomething(SafeHandle h)
{
    IntPtr somethingElse = GetNestedPtr(h);
    // marker 2

    SomethingElse(somethingElse);
}
```

In debug builds `h` isn't subject to finalization until marker 1.  In release builds it's subject to finalization at marker 2.  Aggressive GCStress will force the finalization to happen at/close-to marker 2, revealing the situations where you have a release-only finalizer race condition.

One fix is to put GC.KeepAlive(h) at marker 3.  If you don't own `h` that's still in the toolbox.  In this example, though, it's really the missing `using` which is the bug, and the pending call to Dispose keeps `h` off the finalizer queue, and makes things happen in a predictable order.  A different fix would be to do DangerousAddRef/DangerousRelease on h inside DoSomething(SafeHandle), which keeps it alive during the whole method.

**Comment by karelz** [comment]
I made a note about the GCStress in the new docs - we should include it in advanced scenarios: https://github.com/dotnet/corefx/wiki/New-contributor-Docs#random-useful-stuff

@CIPop can you please try to run it with the GCStress as last attempt to repro in 2.0 timeframe? Thanks!

**Comment by CIPop** [comment]
@karelz Sure, I'll give this a try.

I found this for coreclr: https://github.com/dotnet/coreclr/commit/a97eacb1813191e07b2f9b12b58ecb4ec72bb653

@bartonjs how exactly do I enable this flag for CoreFX? (.Net Framework requires a reg-key change - is it through the coreclr script above?)

**Comment by bartonjs** [comment]
It uses an environment variable.  So:

set COMPlus_GCStress=3
[run the test]

It'll be really, really, really slow :)

**Comment by stephentoub** [comment]
> It'll be really, really, really slow :)

So slow that you probably want to run just that test using corerun explicitly, rather than using msbuild /t:rebuildandtest, etc., as msbuild itself would end up running in that GCStress mode.

**Comment by CIPop** [comment]
Nothing failed... Selectively compiled only the StreamToStream tests and ran them with GCStress:

```
R:\corefx\bin\testhost\netcoreapp-Windows_NT-Debug-x64\dotnet.exe xunit.console.netcore.exe System.Net.Security.Tests.dll -showprogress  -notrait category=failing -notrait category=nonwindowstests
[...]
=== TEST EXECUTION SUMMARY ===
   System.Net.Security.Tests  Total: 52, Errors: 0, Failed: 0, Skipped: 5, Time: 597.738s
```

I'll run the full Security test suite with this enabled next.

**Comment by CIPop** [comment]
Nothing interesting when running the entire test suite:
`   System.Net.Security.Tests  Total: 140, Errors: 0, Failed: 0, Skipped: 7, Time: 855.401s`

I am still waiting for a CI failure on a Debug build which should give us more information on the state of the process.

**Comment by karelz** [comment]
If it didn't repro locally with GCStress (on release), let's move it to Future.

**Comment by CIPop** [comment]
Yesterday's tests were on Debug. Re-ran all tests on Release - nothing failed...

```
R:\corefx\bin\testhost\netcoreapp-Windows_NT-Release-x64\dotnet.exe xunit.console.netcore.exe System.Net.Security.Tests.dll -showprogress  -notrait category=failing -notrait category=nonwindowstests

=== TEST EXECUTION SUMMARY ===
   System.Net.Security.Tests  Total: 140, Errors: 0, Failed: 0, Skipped: 7, Time: 800.419s
```

**Comment by karelz** [comment]
Thanks! Let's leave this one for Future and let's focus on 2.0 bugs now.

**Comment by stephentoub** [comment]
Another failure here:
https://mc.dot.net/#/user/dotnet-bot/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/277eca92322cdd47f32afd206853634f93b79989/workItem/System.Net.Security.Tests/wilogs
```
2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output: FailFast: AcquireCredentialsHandle failed.

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output: System.ComponentModel.Win32Exception (0x80004005): The credentials supplied to the package were not recognized

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.SSPIWrapper.AcquireCredentialsHandle(SSPIInterface secModule, String package, CredentialUse intent, SCHANNEL_CRED scc)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStreamPal.<>c__DisplayClass20_0.<AcquireCredentialsHandle>b__0()

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Security.Principal.WindowsIdentity.<>c__DisplayClass57_0`1.<RunImpersonated>b__0()

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Security.Principal.WindowsIdentity.<>c__DisplayClass61_0.<RunImpersonatedInternal>b__0(Object <state>)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Security.Principal.WindowsIdentity.RunImpersonatedInternal(SafeAccessTokenHandle token, Action action)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Security.Principal.WindowsIdentity.RunImpersonated[T](SafeAccessTokenHandle safeAccessTokenHandle, Func`1 func)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(CredentialUse credUsage, SCHANNEL_CRED secureCredential)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Diagnostics.Debug.Assert(Boolean condition, String message, String detailMessage)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(CredentialUse credUsage, SCHANNEL_CRED secureCredential)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(X509Certificate certificate, SslProtocols protocols, EncryptionPolicy policy, Boolean isServer)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SecureChannel.AcquireServerCredentials(Byte[]& thumbPrint)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SecureChannel.GenerateToken(Byte[] input, Int32 offset, Int32 count, Byte[]& output)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SecureChannel.NextMessage(Byte[] incoming, Int32 offset, Int32 count)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.ForceAuthentication(Boolean receiveFirst, Byte[] buffer, AsyncProtocolRequest asyncRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.ProcessAuthentication(LazyAsyncResult lazyResult)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStream.BeginAuthenticateAsServer(X509Certificate serverCertificate, Boolean clientCertificateRequired, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation, AsyncCallback asyncCallback, Object asyncState)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStream.<>c__DisplayClass35_0.<AuthenticateAsServerAsync>b__0(AsyncCallback callback, Object state)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl(Func`3 beginMethod, Func`2 endFunction, Action`1 endAction, Object state, TaskCreationOptions creationOptions)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.TaskFactory.FromAsync(Func`3 beginMethod, Action`1 endMethod, Object state)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStream.AuthenticateAsServerAsync(X509Certificate serverCertificate, Boolean clientCertificateRequired, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.Tests.DummyTcpServer.OnAccept(Task`1 result)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.Tests.DummyTcpServer.<StartListener>b__8_0(Task`1 t)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ThreadPoolWorkQueue.Dispatch()

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output: 

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Environment.FailFast(System.String, System.Exception)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(CredentialUse, SCHANNEL_CRED)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStreamPal.AcquireCredentialsHandle(System.Security.Cryptography.X509Certificates.X509Certificate, System.Security.Authentication.SslProtocols, System.Net.Security.EncryptionPolicy, Boolean)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SecureChannel.AcquireServerCredentials(Byte[] ByRef)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SecureChannel.GenerateToken(Byte[], Int32, Int32, Byte[] ByRef)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SecureChannel.NextMessage(Byte[], Int32, Int32)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.StartSendBlob(Byte[], Int32, System.Net.AsyncProtocolRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.ProcessReceivedBlob(Byte[], Int32, System.Net.AsyncProtocolRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.StartReadFrame(Byte[], Int32, System.Net.AsyncProtocolRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.StartReceiveBlob(Byte[], System.Net.AsyncProtocolRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.ForceAuthentication(Boolean, Byte[], System.Net.AsyncProtocolRequest)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslState.ProcessAuthentication(System.Net.LazyAsyncResult)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStream.BeginAuthenticateAsServer(System.Security.Cryptography.X509Certificates.X509Certificate, Boolean, System.Security.Authentication.SslProtocols, Boolean, System.AsyncCallback, System.Object)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStream+<>c__DisplayClass35_0.<AuthenticateAsServerAsync>b__0(System.AsyncCallback, System.Object)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.TaskFactory`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].FromAsyncImpl(System.Func`3<System.AsyncCallback,System.Object,System.IAsyncResult>, System.Func`2<System.IAsyncResult,System.Threading.Tasks.VoidTaskResult>, System.Action`1<System.IAsyncResult>, System.Object, System.Threading.Tasks.TaskCreationOptions)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.TaskFactory.FromAsync(System.Func`3<System.AsyncCallback,System.Object,System.IAsyncResult>, System.Action`1<System.IAsyncResult>, System.Object)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.SslStream.AuthenticateAsServerAsync(System.Security.Cryptography.X509Certificates.X509Certificate, Boolean, System.Security.Authentication.SslProtocols, Boolean)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.Tests.DummyTcpServer.OnAccept(System.Threading.Tasks.Task`1<System.Net.Sockets.TcpClient>)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Net.Security.Tests.DummyTcpServer.<StartListener>b__8_0(System.Threading.Tasks.Task`1<System.Net.Sockets.TcpClient>)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)

2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output:    at System.Threading.ThreadPoolWorkQueue.Dispatch()

```

**Comment by CIPop** [comment]
Thanks @stephentoub! The failure is from a Win7 Debug version so it should have generated a dump. 

@mmitche @mellinoe I see the following log at https://mc.dot.net/#/user/dotnet-bot/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/277eca92322cdd47f32afd206853634f93b79989/workItem/System.Net.Security.Tests/wilogs:

```
2017-06-01 14:18:37,890: INFO: proc(54): run_and_log_output: Output: Discovering: System.Net.Security.Tests
2017-06-01 14:18:38,342: INFO: proc(54): run_and_log_output: Output: Discovered:  System.Net.Security.Tests
2017-06-01 14:18:38,437: INFO: proc(54): run_and_log_output: Output: Starting:    System.Net.Security.Tests
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:    System.Net.Security.Tests.SslStreamEKUTest.SslStream_NoEKUServerAuth_Ok [SKIP]
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:       Condition(s) not met: \"IsRootCertificateInstalled\"
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:    System.Net.Security.Tests.SslStreamEKUTest.SslStream_SelfSignedClientEKUClientAuth_Ok [SKIP]
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:       Condition(s) not met: \"IsRootCertificateInstalled\"
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:    System.Net.Security.Tests.SslStreamEKUTest.SslStream_NoEKUClientAuth_Ok [SKIP]
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:       Condition(s) not met: \"IsRootCertificateInstalled\"
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:    System.Net.Security.Tests.SslStreamEKUTest.SslStream_ServerEKUClientAuth_Fails [SKIP]
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:       Condition(s) not met: \"IsRootCertificateInstalled\"
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:    System.Net.Security.Tests.SslStreamEKUTest.SslStream_ClientEKUServerAuth_Fails [SKIP]
2017-06-01 14:18:46,562: INFO: proc(54): run_and_log_output: Output:       Condition(s) not met: \"IsRootCertificateInstalled\"
2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output: FailFast: AcquireCredentialsHandle failed.
2017-06-01 14:18:46,937: INFO: proc(54): run_and_log_output: Output: System.ComponentModel.Win32Exception (0x80004005): The credentials supplied to the package were not recognized
```

but I cannot find the corresponding dump on dumpling. 

Is there any way to get the dump now?

**Comment by mellinoe** [comment]
I'm not sure if Helix runs are configured to use Dumpling -- I don't see it in the logs. @MattGal Do you know if Dumpling is active there?

**Comment by MattGal** [comment]
Yes, I think that's even what Dumpling was originally set up for (Stress runs on Helix).

EnableDumpling is always set if EnableCloudTest is true:

https://github.com/dotnet/buildtools/blob/d69317b58326a7370c729be7e6a6ffd04ffb33bb/src/Microsoft.DotNet.Build.Tasks/PackageFiles/tests.targets#L411 

Skimming an official test build log I don't see any immediate reason why this wouldn't happen, but can help investigate later today if needed.

**Comment by CIPop** [comment]
@MattGal is there any way to obtain the dump from this particular failure now?

**Comment by CIPop** [comment]
We didn't see this issue for the last 2 months. Please reopen adding actionable information (full-memory dump file) or ETW logs if this happens again.
